### PR TITLE
configure entity descriptor based on preset value

### PR DESCRIPTION
**Pull request by flyinghermit**
If entity descriptor is not provided when creating a SAML IdP service provider, we used to first try and fetch the entity descriptor from given Entity ID endpoint. If that would fail, we would generate it ourselves. Now, with the `preset` types, we consider preset value to decide whether to go through fetch and gen process or directly generate entity descriptor. As part of this PR, we now directly generate the entity descriptor for the configured `gcp-workforce` preset type. 

This PR also adds the following:
- `Get` func for for the `Preset` field of `SAMLIdPServiceProviderV1` type. 
- Emit error log for the failed entity descriptor generator.

### CODE REVIEW

File: api/types/saml_idp_service_provider.go [hunk 1]
```diff
@@ -129,6 +129,10 @@ type SAMLIdPServiceProvider interface {
 	GetACSURL() string
 	// SetACSURL sets the ACS URL.
 	SetACSURL(string)
+	// GetPreset returns the Preset.
+	GetPreset() string
+	// SetPreset sets the Preset.
+	SetPreset(string)
```
Comment (smallinsky) [inline api/types/saml_idp_service_provider.go:135]: This seems to be not used and only source of preset is yaml/json marshaling.

File: lib/services/local/saml_idp_service_provider.go [hunk 2]
```diff
@@ -112,15 +113,9 @@ func (s *SAMLIdPServiceProviderService) GetSAMLIdPServiceProvider(ctx context.Co
...
-		if err := s.fetchAndSetEntityDescriptor(sp); err != nil {
-			s.log.Debugf("Failed to fetch entity descriptor from %q. %v.", sp.GetEntityID(), err)
-			// We aren't interested in checking error type as any occurrence of error mean entity descriptor was not set.
-			if err := s.generateAndSetEntityDescriptor(sp); err != nil {
-				return trace.BadParameter("could not generate entity descriptor with given entity_id %q and acs_url %q: %v",
-					sp.GetEntityID(), sp.GetACSURL(), err)
-			}
+		if err := s.configureEntityDescriptorPerPreset(sp); err != nil {
+			return trace.Wrap(trace.BadParameter("failed to configure entity descriptor with the given entity_id %q and acs_url %q: %v",
+				sp.GetEntityID(), sp.GetACSURL(), err))
```
Comment (smallinsky) [inline lib/services/local/saml_idp_service_provider.go:118]: I don't thing that you need to use `trace.Wrap` to additionally wrap the `trace.BadParameter`
```suggestion
			return trace.BadParameter("failed to configure entity descriptor with the given entity_id %q and acs_url %q: %v",
				sp.GetEntityID(), sp.GetACSURL(), err)
```

File: lib/services/local/saml_idp_service_provider.go [hunk 3]
```diff
@@ -243,6 +238,25 @@ func (s *SAMLIdPServiceProviderService) ensureEntityIDIsUnique(ctx context.Conte
 	return nil
 }
 
+// configureEntityDescriptorPerPreset configures entity descriptor based on SAML service provider preset.
+func (s *SAMLIdPServiceProviderService) configureEntityDescriptorPerPreset(sp types.SAMLIdPServiceProvider) error {
+	switch sp.GetPreset() {
+	case preset.GCPWorkforce:
+		return s.generateAndSetEntityDescriptor(sp)
```
Comment (smallinsky) [inline lib/services/local/saml_idp_service_provider.go:245]:
```suggestion
		return trace.Wrap(s.generateAndSetEntityDescriptor(sp))
```

File: lib/services/local/saml_idp_service_provider.go [hunk 4]
```diff
@@ -243,6 +238,25 @@ func (s *SAMLIdPServiceProviderService) ensureEntityIDIsUnique(ctx context.Conte
...
+	default:
+		// fetchAndSetEntityDescriptor is expected to return error if it fails
+		// to fetch a valid entity descriptor.
+		if err := s.fetchAndSetEntityDescriptor(sp); err != nil {
+			s.log.Debugf("Failed to fetch entity descriptor from %q. %v.", sp.GetEntityID(), err)
+			// We aren't interested in checking error type as any occurrence of error
+			// mean entity descriptor was not set.
+			return s.generateAndSetEntityDescriptor(sp)
```
Comment (smallinsky) [inline lib/services/local/saml_idp_service_provider.go:253]:
```suggestion
			return trace.Wrap(s.generateAndSetEntityDescriptor(sp))
```

File: lib/services/local/saml_idp_service_provider.go [hunk 5]
```diff
@@ -243,6 +238,25 @@ func (s *SAMLIdPServiceProviderService) ensureEntityIDIsUnique(ctx context.Conte
...
+// configureEntityDescriptorPerPreset configures entity descriptor based on SAML service provider preset.
+func (s *SAMLIdPServiceProviderService) configureEntityDescriptorPerPreset(sp types.SAMLIdPServiceProvider) error {
+	switch sp.GetPreset() {
+	case preset.GCPWorkforce:
+		return s.generateAndSetEntityDescriptor(sp)
+	default:
+		// fetchAndSetEntityDescriptor is expected to return error if it fails
+		// to fetch a valid entity descriptor.
+		if err := s.fetchAndSetEntityDescriptor(sp); err != nil {
+			s.log.Debugf("Failed to fetch entity descriptor from %q. %v.", sp.GetEntityID(), err)
```
Comment (smallinsky) [inline lib/services/local/saml_idp_service_provider.go:250]: typo `.`  -> `:` ? 
```suggestion
			s.log.Debugf("Failed to fetch entity descriptor from %q: %v.", sp.GetEntityID(), err)
```

**Review by smallinsky — APPROVED** [review]

File: lib/services/local/saml_idp_service_provider.go (hunk 2 above)
Comment (flyinghermit) [reply lib/services/local/saml_idp_service_provider.go]: yaa wrapping it below like you suggested in https://github.com/gravitational/teleport/pull/39385#discussion_r1525921803 seems like more proper place.

File: api/types/saml_idp_service_provider.go (hunk 1 above)
Comment (flyinghermit) [reply api/types/saml_idp_service_provider.go]: Good point, removed in https://github.com/gravitational/teleport/pull/39385/commits/a7788f0f7f6e9890ff705ce6c2f73f11d8a55b2d

**Comment by flyinghermit** [comment]
Friendly ping @ibeckermayer

**Review by atburke — APPROVED** [review]

**Pull request merged**